### PR TITLE
Xpetra: Fix Issue 5672 and ETI Xpetra_MapFactory

### LINKED ISSUE #5672: Xpetra: Compile error in BlockedMap
@trilinos/xpetra 

## Description
Kevin Cox sent us a report of a compilation he was getting that looks like it might be something from the Xpetra ETI work.  I put together a configuration that is as close as I can get to the one he reported but it does not replicate his error.  The biggest difference might be that he's on Geminga but I don't have an account there.  He's also loading AmgX and I don't have that in my environment.

### Error Message
Here's the error message he's seeing:
```
/home/kcox/Trilinos/packages/xpetra/src/BlockedMap/Xpetra_BlockedMap.hpp(537): error: incomplete type is not allowed
          detected during:
            instantiation of "void Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node>::assign(const Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node> &) [with LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_dyn_cast.hpp(175): here
            instantiation of "T_To &Teuchos::dyn_cast<T_To,T_From>(T_From &) [with T_To=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T_From=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_RCP.hpp(768): here
            instantiation of "Teuchos::RCP<T2> Teuchos::rcp_dynamic_cast<T2,T1>(const Teuchos::RCP<T1> &, bool) [with T2=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T1=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/xpetra/src/MultiVector/Xpetra_MultiVectorFactory_def.hpp(68): here
            instantiation of "Teuchos::RCP<Xpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>> Xpetra::MultiVectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(const Teuchos::RCP<const Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, size_t, bool) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/xpetra/src/Utils/Xpetra_ETI_4arg_mv.hpp(42): here
 
/home/kcox/Trilinos/packages/xpetra/src/BlockedMap/Xpetra_BlockedMap.hpp(543): error: incomplete type is not allowed
          detected during:
            instantiation of "void Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node>::assign(const Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node> &) [with LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_dyn_cast.hpp(175): here
            instantiation of "T_To &Teuchos::dyn_cast<T_To,T_From>(T_From &) [with T_To=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T_From=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_RCP.hpp(768): here
            instantiation of "Teuchos::RCP<T2> Teuchos::rcp_dynamic_cast<T2,T1>(const Teuchos::RCP<T1> &, bool) [with T2=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T1=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/xpetra/src/MultiVector/Xpetra_MultiVectorFactory_def.hpp(68): here
            instantiation of "Teuchos::RCP<Xpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>> Xpetra::MultiVectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(const Teuchos::RCP<const Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, size_t, bool) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/xpetra/src/Utils/Xpetra_ETI_4arg_mv.hpp(42): here
 
/home/kcox/Trilinos/packages/xpetra/src/BlockedMap/Xpetra_BlockedMap.hpp(545): error: incomplete type is not allowed
          detected during:
            instantiation of "void Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node>::assign(const Xpetra::BlockedMap<LocalOrdinal, GlobalOrdinal, Node> &) [with LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_dyn_cast.hpp(175): here
            instantiation of "T_To &Teuchos::dyn_cast<T_To,T_From>(T_From &) [with T_To=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T_From=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/teuchos/core/src/Teuchos_RCP.hpp(768): here
            instantiation of "Teuchos::RCP<T2> Teuchos::rcp_dynamic_cast<T2,T1>(const Teuchos::RCP<T1> &, bool) [with T2=const Xpetra::BlockedMap<int, int, Kokkos_Compat_KokkosCudaWrapperNode>, T1=const Xpetra::Map<int, Xpetra::Details::DefaultGlobalOrdinal, KokkosClassic::DefaultNode::DefaultNodeType>]"
/home/kcox/Trilinos/packages/xpetra/src/MultiVector/Xpetra_MultiVectorFactory_def.hpp(68): here
            instantiation of "Teuchos::RCP<Xpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>> Xpetra::MultiVectorFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(const Teuchos::RCP<const Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> &, size_t, bool) [with Scalar=double, LocalOrdinal=int, GlobalOrdinal=int, Node=Kokkos_Compat_KokkosCudaWrapperNode]"
/home/kcox/Trilinos/packages/xpetra/src/Utils/Xpetra_ETI_4arg_mv.hpp(42): here
 
3 errors detected in the compilation of "/tmp/tmpxft_000099b0_00000000-5_Xpetra_MultiVectorFactory.cpp4.ii".
make[2]: *** [packages/xpetra/src/CMakeFiles/xpetra.dir/Utils/ExplicitInstantiation/Xpetra_MultiVectorFactory.cpp.o] Error 2
make[1]: *** [packages/xpetra/src/CMakeFiles/xpetra.dir/all] Error 2
make: *** [all] Error 2
```

#### Configuration
This is the configuration Kevin sent:
```
rm -rf CMakeCache.txt CMakeFiles cmake
 
cmake \
-D Trilinos_ENABLE_Epetra=OFF \
-D Trilinos_ENABLE_Kokkos=ON \
-D KOKKOS_ARCH="SNB;Kepler35" \
-D Kokkos_ENABLE_Cuda=ON \
-D Kokkos_ENABLE_Cuda_UVM=ON \
-D Kokkos_ENABLE_Cuda_Lambda=ON \
-D Ifpack2_ENABLE_TESTS=ON \
-D Ifpack2_ENABLE_EXAMPLES=ON \
-D MueLu_ENABLE_TESTS=ON \
-D Trilinos_ENABLE_MueLu=ON \
-D Trilinos_ENABLE_Pamgen=OFF \
-D Trilinos_ENABLE_ShyLU_Node=OFF \
-D Trilinos_ENABLE_Tpetra=ON \
-D Trilinos_ENABLE_Xpetra=ON \
-D Tpetra_ENABLE_TESTS=ON \
-D Tpetra_ENABLE_EXAMPLES=ON \
  -D Tpetra_INST_INT_LONG=OFF \
  -D Tpetra_INST_INT_INT=ON \
  -D Tpetra_INST_INT_LONG_LONG=OFF \
\
-D Trilinos_ENABLE_Zoltan=OFF \
-D TPL_ENABLE_CUDA=ON \
-D TPL_ENABLE_MPI=ON \
\
-DTPL_ENABLE_AmgX=ON \
-D TPL_AmgX_LIBRARIES=/home/kcox/Trilinos/AMGX/build/install/lib/libamgxsh.so \
-DAmgX_LIBRARY_DIRS=/home/kcox/Trilinos/AMGX/build/install/lib/ \
-DAmgX_INCLUDE_DIRS=/home/kcox/Trilinos/AMGX/build/install/include \
-D TPL_ENABLE_CUSPARSE=ON \
\
-D Trilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
\
\ #-D CMAKE_C_FLAGS="-g" \
\ #-D CMAKE_CXX_FLAGS="-g" \
..
```

FYI @csiefer2

### PR DESCRIPTION
**Pull request by william76**
@trilinos/xpetra 

Fixes #5672 

The fix looks like it required an ETI of MapFactory.  There were a couple of other small bugs I located and fixed with this too.

### CODE REVIEW

**Comment by william76** [comment]
I'm running a test on Geminga to test out the build that's reported in #5672 which is still running so I'm not setting Automerge until I can review that build... but in the meantime, let's let the autotester do its thing too.

File: packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_decl.hpp [hunk 1]
```diff
@@ -115,7 +119,7 @@ namespace Xpetra {
      *
      * \param map BlockedMap defining the block structure of the multi vector
      */
-    MapExtractor(const Teuchos::RCP< const BlockedMap > &map);
+    MapExtractor(const Teuchos::RCP< const Xpetra::BlockedMap<LocalOrdinal,GlobalOrdinal,Node> > &map);
```
Comment (mhoemmen) [inline packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_decl.hpp]: You don't need to leave a space between the two closing angle brackets any more (with C++11).

File: packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+         class Node = typename Map<LocalOrdinal, GlobalOrdinal>::node_type>
+class MapFactory
+{
+
+
+  private:
+
+
+    //! Private constructor. This is a static class.
+    MapFactory() {}
```
Comment (mhoemmen) [inline packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp]: The correct C++11 idiom is 
```c++
MapFactory() = delete;
```
This does not need to be `private`.  The correct C++98 idiom is to declare the constructor private, without defining (implementing) it.

File: packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+
+
+  public:
+
+
+    //! Map constructor with Xpetra-defined contiguous uniform distribution.
+#ifdef TPETRA_ENABLE_DEPRECATED_CODE
```
Comment (mhoemmen) [inline packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp]: This is an Xpetra thing, but you're using a Tpetra macro here.  Is that OK?  What if Tpetra is not enabled?

Comment (william76) [reply packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp]: @mhoemmen the `TPETRA_ENABLE_DEPRECATED_CODE` stuff isn't mine... that was already in develop when I forked my branch @tjfulle ?

File: packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp (hunk 2 above)
Comment (william76) [reply packages/xpetra/src/Map/Xpetra_MapFactory_decl.hpp]: @mhoemmen this is how the code was before I worked on ETI.  I will not change the private / public / protected status of anything in these classes for ETI.

File: packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_decl.hpp (hunk 1 above)
Comment (william76) [reply packages/xpetra/src/BlockedCrsMatrix/Xpetra_MapExtractor_decl.hpp]: @mhoemmen I know. In some places I'm changing things to the C++11 style but there's a lot of code stuff in these ETI functions that I'm not really touching so it's just staying how it was in the original, pre-ETI code.

**Comment by william76** [comment]
@csiefer2 ok, so this got past gcc 4.8.4 but what I did is kludgy.  If we want to merge it in, I'll want to do a cleanup pass to remove some scaffolding comments, etc.

Here's the function that I modified:

https://github.com/trilinos/Trilinos/blob/22180caf8b3aba38c362e3b758b12af4d5889da6/packages/xpetra/src/Map/Xpetra_MapFactory_def.hpp#L258-L285

The problem was the code on line 276:
https://github.com/trilinos/Trilinos/blob/22180caf8b3aba38c362e3b758b12af4d5889da6/packages/xpetra/src/Map/Xpetra_MapFactory_def.hpp#L276

GCC 4.8.4 didn't like this but newer compilers seemed fine with it.  From what I can tell in the errors that CDash was throwing out we were trying to assign `Tpetra::Map` objects that had different Node types (one was OpenMP and the other is Serial)... I'm not sure if a deeper fix needs to be made in `Tpetra::createLocalMap` which is an external c'tor for `Tpetra::Map` that takes just LO and GO and uses the default `Kokkos::Node` type but `Xpetra::MapFactory` is LO-GO-NO and has a default option.

I added the guard based on GCC version to use the code that was existing already in `MapFactory`  if we're GCC 5.x.x or higher and throw an error otherwise.  I expected that this would get thrown by the GCC 4.8.4 tests but since the tests all apparently passed PR testing, can we say that `Xpetra::MapFactory::createLocalMap` is never actually getting tested in the Trilinos test suite?

Going forward, do we cleanup/merge this in?  Should I test out another hack that if we're GCC > 5.x then instead of using `Tpetra::createLocalMap` we try `Tpetra::createLocalMapWithNode` to explicitly match the node type?

I'll add the WIP flag for now and we can discuss Monday.

**Comment by mhoemmen** [comment]
@william76 wrote:
> From what I can tell in the errors that CDash was throwing out we were trying to assign `Tpetra::Map` objects that had different Node types (one was OpenMP and the other is Serial)....

`Tpetra::createLocalMap<LocalOrdinal,GlobalOrdinal>(...)` returns a `Tpetra::Map` with Tpetra's default Node type.  That could differ from `Node` in `Xpetra::TpetraMap<LocalOrdinal,GlobalOrdinal,Node>`.

Fix this by using the two-argument overload of `Tpetra::createLocalMapWithNode`.  Give that function all three template arguments (including the same `Node` type that you give to `Xpetra::TpetraMap`).

**Comment by william76** [comment]
@mhoemmen 
There is a `Xpetra::MapFactory::createLocalMapWithNode()` version already that uses that function and the `Xpetra::MapFactory::createLocalMap()` which has the problem, apparently in the gcc 4.8.4 build but not newer compilers (??).  Are you saying that in `Xpetra::MapFactory::createLocalMap()` I should replace its call to `Tpetra::Map::createLocalMap()` with a call to `Tpetra::Map::createLocalMapWithNode` and use LocalOrdinal, GlobalOrdinal, and Node?

**Comment by csiefer2** [comment]
I believe @mhoemmen right.  That should fix the problem.

Add this to the list of "code nobody ever build which ETI exposed as an issue"

Ugh.

**Comment by lucbv** [comment]
@william76  @csiefer2 this code is actually used in some of the refactored factories (TentativePFactory_kokkos at least), it essentially allows us to create maps using `Kokkos::View` instead of `Teuchos::Array`.
I would assume that we are testing the refactore TentaviePFactory somewhere so it is strange that we did not catch this somehow?

**Comment by csiefer2** [comment]
@lucbv I suspect we never turned multiple node types on at the same time during testing...

**Comment by lucbv** [comment]
@csiefer2 I see, maybe we could have a `serial` + `openMP` build to test that scenario?

**Pull request merged**